src/App.tsx

```diff
@@ -15,36 +17,106 @@ export const goodsFromServer = [
   'Garlic',
 ];
 
+enum SortType {
+  Alpha = 'alpha',
+  Length = 'length',
+  Default = 'default',
+}
+
+type SortingFunction = (a: string, b: string) => number;
+
+const SORTING_FUNCTIONS: { [key in SortType]: SortingFunction } = {
+  [SortType.Alpha]: (a, b) => a.localeCompare(b),
+  [SortType.Length]: (a, b) => a.length - b.length,
+  [SortType.Default]: () => 0,
+};
+
 export const App: React.FC = () => {
+  const [visibleGoods, setVisibleGoods] = useState<string[]>(goodsFromServer);
+  const [sortField, setSortField] = useState<SortType>(SortType.Default);
+  const [isReversed, setIsReversed] = useState<boolean>(false);
+
+  const handleGoods = () => {
+    const newGoods = [...goodsFromServer];
+
+    if (sortField !== SortType.Default) {
+      newGoods.sort(SORTING_FUNCTIONS[sortField]);
+    }
+
+    return isReversed
+      ? setVisibleGoods(newGoods.reverse())
+      : setVisibleGoods(newGoods);
+  };
+
+  const sortBy = (type: SortType) => {
+    setSortField(type);
+  };
+
+  const reverse = () => {
+    setIsReversed(r => !r);
+  };
+
+  const reset = () => {
+    setSortField(SortType.Default);
+    setIsReversed(false);
+  };
+
+  const isOriginalList = () => {
+    return !isReversed && sortField === SortType.Default;
+  };
+
+  useEffect(() => {
+    handleGoods();
+  }, [sortField, isReversed]);
+
   return (
     <div className="section content">
       <div className="buttons">
-        <button type="button" className="button is-info is-light">
+        <button
+          type="button"
+          onClick={() => sortBy(SortType.Alpha)}
+          className={cn('button is-info', {
+            'is-light': sortField !== SortType.Alpha,
+          })}
+        >
           Sort alphabetically
         </button>
 
-        <button type="button" className="button is-success is-light">
+        <button
+          type="button"
+          onClick={() => sortBy(SortType.Length)}
```

Comment:
a better practice would be to add a handler for each SortType - in general you should always have plain functions passed to JS events to improve mantainability